ojd_daps_skills/utils/duplication.py

```diff
@@ -0,0 +1,120 @@
+"""
+Functions to deal with duplication in job adverts.
+
+get_deduplicated_job_adverts takes the job adverts (only job id and date columns are needed)
+and chunks up the dates and then removes duplicates from each chunk, before recombining.
+
+The duplicates output from the OJO pipeline is needed here (which is a list of pairs of job ids
+which are semantically the same).
+"""
+
+from datetime import datetime, timedelta, date
+from dateutil.relativedelta import relativedelta
+
+
+def get_date_ranges(start, end, num_units=7, unit_type="days"):
+    """
+    Chunk up a range of dates into user specified units of time.
+    Inputs:
+        start, end: strings of a date in the form %Y-%m-%d (e.g. '2021-05-31')
+        num_units: time period of chunks, e.g. if num_units=3 and unit_type="days" then this is 3 days
+        unit_type: ["days", "weeks", "months"]
+    Outputs:
+        A list of the start and end dates to each chunk
+        The last chunk may be partial.
+    """
+    start = datetime.strptime(start, "%Y-%m-%d")
+    end = datetime.strptime(end, "%Y-%m-%d")
+
+    if unit_type == "days":
+        time_addition = timedelta(days=num_units)
+    elif unit_type == "weeks":
+        time_addition = timedelta(weeks=num_units)
+    elif unit_type == "months":
+        time_addition = relativedelta(months=num_units)
+    else:
+        print('unit_type not recognised - use ["days", "weeks", "months"]')
+
+    chunk_start = start
+    chunk_end = chunk_start + time_addition
+    list_ranges = []
+    while chunk_end < end:
+        list_ranges.append((chunk_start, chunk_end))
+        chunk_start += time_addition
+        chunk_end = chunk_start + time_addition
+    list_ranges.append((chunk_start, end))
+    return list_ranges
+
+
+def create_date_chunks_mapper(first_date, last_date, num_units=7, unit_type="days"):
+    """
+    Gets a dict of all dates in the range and which start date chunk they map to.
+    e.g. if num_units=2 and unit_type="days"
+    {
+    '2021-06-01': '2021-06-01',
+    '2021-06-02': '2021-06-01',
+    '2021-06-03': '2021-06-03',
+    '2021-06-04': '2021-06-03'
+    }
+    """
+
+    date_ranges_list = get_date_ranges(
+        first_date, last_date, num_units, unit_type=unit_type
+    )
+
+    date_chunk_mapper = {}
+    for start, end in date_ranges_list:
+        curr_date = start
+        while curr_date < end:
+            date_chunk_mapper[curr_date.strftime("%Y-%m-%d")] = start.strftime(
+                "%Y-%m-%d"
+            )
+            curr_date += timedelta(days=1)
+        date_chunk_mapper[curr_date.strftime("%Y-%m-%d")] = start.strftime("%Y-%m-%d")
+    return date_chunk_mapper
+
+
+def get_deduplicated_job_adverts(
```

Comment:
@Jack-Vines this is how I've dealt with duplication (using s3://open-jobs-lake/escoe_extension/outputs/data/model_application_data/raw_job_adverts_sample.csv and s3://open-jobs-lake/escoe_extension/outputs/data/model_application_data/job_ad_duplicates_sample.csv) 
Please could you check it over and I'm happy to chat if anything is unclear?

FYI @india-kerle